### PR TITLE
odb SQLite + odb compiler + example

### PR DESCRIPTION
**Pull request by pretyman**
Here are the packages for C++ Object-Relational Mapping (ORM) odb - http://www.codesynthesis.com/products/odb/

This commit contains:
* SQLite3 with autotools and flags -DSQLITE_ENABLE_UNLOCK_NOTIFY=1 -DSQLITE_ENABLE_RTREE=1 (maybe this package is unwanted by hunter, since amalgamation of SQLite3 is usually preferred)
* odb core runtime
* odb compiler
* odb-sqlite for persisting in SQLite3 databases 
* examples/odb-sqlite with a working example of the compiler command
* find scripts for SQLite3, odb and odb-compiler

### CODE REVIEW

**Comment by pretyman** [comment]
I would love to make odb cacheable, but if I do it, and later I try to compile odb-sqlite, which depends on it, ODB_ROOT is pointing to the build directory and not at the final destination. I looked around in hunter code to see how I could get around this, but didn't find a concrete solution. 

Hence the call hunter_cacheable(odb) is commented for now.

If these packages are not interesting for odb, I can host them on my fork without any problems, but would be cool to see them in the upstream.

Tested on the Mac platform only.

**Comment by pretyman** [comment]
I forgot to mention: SQLite3 is compiled with those options so then SpatiaLite will be able to be built with it. I should push SpatiaLite (and its dependencies) this upcoming week.

File: cmake/find/FindSQLite3.cmake [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# ----------------------------------------------------------------------
+# FindSQLite3.cmake - SQLite3 CMake module
+#
+# Copyright (C) 2015 Alexandre Pretyman. All rights reserved.
+#
+# ----------------------------------------------------------------------
+
+if(HUNTER_STATUS_DEBUG)
+  message("[hunter] SQLITE3_ROOT ${SQLITE3_ROOT}")
```
Comment (ruslo) [inline cmake/find/FindSQLite3.cmake]: You can use `hunter_status_debug` here

File: cmake/find/FindSQLite3.cmake [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    set(SQLITE3_FOUND TRUE)
+    add_library("SQLite3" UNKNOWN IMPORTED)
+    set_target_properties(
+        "SQLite3"
+        PROPERTIES
+        INTERFACE_INCLUDE_DIRECTORIES "${SQLITE3_INCLUDE_DIR}"
+        IMPORTED_LOCATION "${SQLITE3_LIBRARY}"
+    )
+  get_filename_component(SQLITE3_LIB_DIR "${SQLITE3_LIBRARY}" DIRECTORY CACHE)
+elseif(SQLite3_FIND_REQUIRED)
```
Comment (ruslo) [inline cmake/find/FindSQLite3.cmake]: +2 spaces

File: cmake/find/FindSQLite3.cmake (hunk 1 above)
Comment (pretyman) [inline cmake/find/FindSQLite3.cmake]: Ok, will do, I was thinking it could be useful outside hunter, but maybe there is no point in that line of thought :-)

File: cmake/find/Findodb-compiler.cmake [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  set_target_properties(${_target} 
+      PROPERTIES
+        SOURCES "${_target_sources}"
+  )
+  target_include_directories(${_target}
+      PUBLIC
+         ${ODB_COMPILE_OUTPUT_DIR}
+        ${ODB_INCLUDE_DIRS}
+  )  
+#set(${outvar} ${${outvar}} PARENT_SCOPE)
```
Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: Please remove this one if it's not needed.

File: cmake/find/Findodb-compiler.cmake [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  endforeach()
+
+  foreach(dir ${PARAM_INCLUDE} ${ODB_INCLUDE_DIRS})
+    list(APPEND ODB_ARGS "-I${dir}")
+  endforeach()
+
+  file(REMOVE_RECURSE "${ODB_COMPILE_OUTPUT_DIR}")
+  file(MAKE_DIRECTORY "${ODB_COMPILE_OUTPUT_DIR}")
+
+  #  list(APPEND ODB_ARGS --odb-epilogue "\#include <stdio")
```
Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: remove?

File: cmake/find/Findodb-compiler.cmake (hunk 3 above)
Comment (pretyman) [inline cmake/find/Findodb-compiler.cmake]: ok, sorry, it slipped :S I will correct the funny extra space too

File: cmake/projects/Android-SDK-Tools/hunter.cmake [hunk 5]
```diff
@@ -62,6 +62,17 @@ elseif(CMAKE_HOST_UNIX)
...
+      URL
+      "http://dl-ssl.google.com/android/repository/tools_r24.3-linux.zip"
+      SHA1
+      ecb86253e97cd14459888701fc3478b4d153a86c
+  )
+
+  hunter_add_version(
+      PACKAGE_NAME
+      Android-SDK-Tools
+      VERSION
```
Comment (ruslo) [inline cmake/projects/Android-SDK-Tools/hunter.cmake]: I guess this code is not related to the change at all :)

Comment (pretyman) [inline cmake/projects/Android-SDK-Tools/hunter.cmake]: Yeah, I can't really explain how this got there, I used the "squashing" instructions from last time to generate the commit, somehow this pushed through...

**Comment by pretyman** [comment]
I will improve the code as per the comments and issue another push during this week.

Thanks for the code review!

File: cmake/find/FindSQLite3.cmake (hunk 1 above)
Comment (ruslo) [inline cmake/find/FindSQLite3.cmake]: Hm... I see. You can add this code:
```cmake
if(NOT HUNTER_ENABLED)
  macro(hunter_status_debug)
    message(${ARGV})
  endmacro()
endif()

File: cmake/projects/SQLite3/hunter.cmake [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+else()
+  set(HUNTER_CMAKE_PROJECTS_SQLITE3_HUNTER_CMAKE 1)
+endif()
+
+include(hunter_add_version)
+include(hunter_cacheable)
+include(hunter_download)
+include(hunter_pick_scheme)
+
+# Disable searching in locations not specified by these hint variables.
```
Comment (ruslo) [inline cmake/projects/SQLite3/hunter.cmake]: I think this comment not needed (?)

File: cmake/projects/SQLite3/hunter.cmake [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2013, Ruslan Baratov
```
Comment (ruslo) [inline cmake/projects/SQLite3/hunter.cmake]: Please add yourself here and `s,2013,2015,g'

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+endif()
+
+list(APPEND CMAKE_MODULE_PATH "@HUNTER_SELF@/cmake/find")
+
+set(ODB_ROOT "@ODB_ROOT@")
+find_package(odb REQUIRED)
+
+set(SQLITE3_ROOT "@SQLITE3_ROOT@")
+find_package(SQLite3 REQUIRED)
+
```
Comment (ruslo) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: Do you have any ideas about unifying this scheme. I think if there will be some updates in future this code will be hard to maintain. I.e. on change `+set(ldflags "")` you have to modify all `url_sha1_*_autotools`.

**Comment by ruslo** [comment]
> I would love to make odb cacheable, but if I do it, and later I try to compile odb-sqlite, which depends on it, 

Can you add a simple example with `odb`? If it will work with `hunter_cacheable(odb)` it will mean that `odb` is fine and problem is elsewhere.

> ODB_ROOT is pointing to the build directory and not at the final destination

After command `hunter_add_package(odb)` I expect that this variable points to final destination.

Also if you want to have cacheable packages that depends on something and you don't want to "Hunterize" them you need to add dependencies explicitly. See `PACKAGE_DEPENDS_ON` option in `LLVM`: https://github.com/ruslo/hunter/blob/master/cmake/projects/LLVM/hunter.cmake

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in (hunk 8 above)
Comment (pretyman) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: quite frankly, it is already a hassle with 3 `url_sha1_*_autotools`. I was thinking of either a macro or function. Something like:

* hunter_autotools_init() - maybe with optional preprocessor flags, {c,cxx} compile flags and linker flags as parameters

I can't figure out what is better, a macro that sets the `configure_opts` with no scope restrictions or a function that sets `configure_opts` on `PARENT_SCOPE`, maybe a function is safer, right? What do you think?

**Comment by pretyman** [comment]
Actually, there is already an example on `examples/odb-sqlite` all is needed is to mkdir build, and run cmake from there.

I updated it to use `PACKAGE_DEPENDS_ON`, let's see how will it work. I will get back to this thread when I test it.

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in (hunk 8 above)
Comment (pretyman) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: I think I got Chip and Dale (my two neurons) busy working on this. 

The API that would be best would be a *function* like:
```cmake
hunter_autotools_configure_options(configure_opts
    CPPFLAGS "-DEXAMPLE -I/usr/local/this/and/that"
    CXXFLAGS "-extra-stuff"
    CFLAGS "-other-flags"
    LDFLAGS "-L/usr/local/lib -llibrary"
    EXTRAFLAGS "--enable-feature"
)
```

`hunter_autotools_configure_options(...)` would add the toolchain binaries and cross compiling host to `configure_opts`
 
@ruslo  what I need help deciding is:
1. let be the responsibility of the caller to add `${CMAKE_CXX_FLAGS}`.
2. `${CMAKE_CXX_FLAGS}` gets added automatically from inside the function.

`LDFLAGS` is more tricky because depending what the autotools project generates, it could have different linker flags, or so I believe, but I could be very wrong, since I haven't generated a shared object in years, not sure how different it is from static library.

Comment (pretyman) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: I think it makes sense that `${CMAKE_CXX_FLAGS}` gets added by default by the function, leaving the user just the responsibility to add flags other than the default ones and the `LDFLAGS`.

Would it be wise to have `LDFLAGS` set with `${CMAKE_EXE_LINKER_FLAGS}`? Can you think of a use case that would not be desired?

Thanks

**Comment by pretyman** [comment]
I changed all the hunter_autotools* functions for one to be used like this:

```cmake
hunter_autotools_configure_options(configure_opts
    CPPFLAGS                        # C pre-processor flags
      "-DEXAMPLE -I/usr/local/include/library"
    CFLAGS                          # C Compiler flags
      "-super-optimation"
    CXXFLAGS                        # CXX compiler flags
      "-super-optimation"
    LDFLAGS                          # Linker flags
      "-L/usr/local/lib -llibrary"
    EXTRA_FLAGS                      # any other flag that will go at the end
      --enable-feature
      --disable-other
      --with-library
)
```
I documented Findodb-compiler.cmake a littke

If you want I can do the squash commit, if you find helpful.

Thanks

File: cmake/find/FindSQLite3.cmake [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+#
+# ----------------------------------------------------------------------
+
+if(NOT HUNTER_ENABLED)
+  macro(hunter_status_debug)
+    message(${ARGV})
+  endmacro()
+endif()
+
+hunter_status_debug("[hunter] SQLITE3_ROOT ${SQLITE3_ROOT}")
```
Comment (ruslo) [inline cmake/find/FindSQLite3.cmake]: `[hunter]` not needed since it will be added by `hunter_status_debug`

File: cmake/find/FindSQLite3.cmake [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  if(SQLITE3_INCLUDE_DIR AND SQLITE3_LIBRARY)
+    set(SQLITE3_FOUND TRUE)
+    add_library("SQLite3" UNKNOWN IMPORTED)
+    set_target_properties("SQLite3"
+        PROPERTIES
+          INTERFACE_INCLUDE_DIRECTORIES "${SQLITE3_INCLUDE_DIR}"
+          IMPORTED_LOCATION "${SQLITE3_LIBRARY}"
+    )
+  get_filename_component(SQLITE3_LIB_DIR "${SQLITE3_LIBRARY}" DIRECTORY CACHE)
+elseif(SQLite3_FIND_REQUIRED)
```
Comment (ruslo) [inline cmake/find/FindSQLite3.cmake]: +2 space

File: cmake/find/Findodb-compiler.cmake [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+
+if(NOT HUNTER_ENABLED)
+  macro(hunter_status_debug)
+    message(${ARGV})
+  endmacro()
+endif()
+
+if(ANDROID) 
+  #AWP: this seems to be only needed by the Android toolchain, maybe this should
+  #     be elsewhere?
```
Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: I think this should work without additional code if program will be installed by regular `install` command. Since `url_sha1_unpack` scheme used this package will be unpacked to `<root-id>/Build/<package>/Source` instead of `<root-id>/Install` so you need to do additional tricks to make `find_program` work. Also probably adding new scheme `url_sha1_install_binaries` is an appropriate alternative.

File: cmake/find/Findodb-compiler.cmake [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+      string(REGEX REPLACE "^.*:" "" sfx "${sfx}")
+
+      if(NOT "${PARAM_MULTI_DATABASE}" MATCHES "static" OR NOT "${pfx}" MATCHES "common")
+        set(output "${ODB_COMPILE_OUTPUT_DIR}/${fname}${sfx}${ODB_COMPILE_SOURCE_SUFFIX}")
+        list(APPEND outputs "${output}")
+        list(APPEND generated_source_files "${output}")
+      endif()
+    endforeach()
+
+    if(HUNTER_STATUS_DEBUG)
```
Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: `hunter_status_debug` can be used

File: cmake/find/Findodb.cmake [hunk 13]
```diff
@@ -0,0 +1,108 @@
+# ----------------------------------------------------------------------
```
Comment (ruslo) [inline cmake/find/Findodb.cmake]: As far as I can see we already have `odb` package in Hunter so this change can be made separately. Can you please open pull request only with this module and simple example (just `hunter_add_package(odb)` + `find_package(odb REQUIRED)`). Also I see no problem with caching this package - what error do you have exactly? Maybe this package is not relocatable by design, i.e. it hardcoding some path while installing? RPATH?

File: cmake/projects/Android-SDK-Tools/hunter.cmake (hunk 5 above)
Comment (ruslo) [inline cmake/projects/Android-SDK-Tools/hunter.cmake]: Still see this one.

File: cmake/modules/hunter_autotools_configure_options.cmake [hunk 14]
```diff
@@ -0,0 +1,176 @@
+# Copyright (c) 2015 Ruslan Baratov, Alexandre Pretyman
```
Comment (ruslo) [inline cmake/modules/hunter_autotools_configure_options.cmake]: Same here. I think this is quite big change so it can be separate pull request for easy review reason.

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+set(HUNTER_STATUS_DEBUG "@HUNTER_STATUS_DEBUG@")
+
+if("@MSVC@")
+  hunter_fatal_error("Autotools scheme not supported with Visual Studio")   
+endif()
+
+list(APPEND CMAKE_MODULE_PATH "@HUNTER_SELF@/cmake/find")
+
+set(ODB_ROOT "@ODB_ROOT@")
```
Comment (ruslo) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: You can just include master file `cmake/Hunter` and use `hunter_add_package(odb)`.

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+hunter_test_string_not_empty("@HUNTER_EP_NAME@")
+hunter_test_string_not_empty("@HUNTER_PACKAGE_URL@")
+hunter_test_string_not_empty("@HUNTER_PACKAGE_SHA1@")
+hunter_test_string_not_empty("@HUNTER_PACKAGE_DOWNLOAD_DIR@")
+hunter_test_string_not_empty("@HUNTER_PACKAGE_SOURCE_DIR@")
+hunter_test_string_not_empty("@HUNTER_PACKAGE_INSTALL_PREFIX@")
+
+set(configure_command "./configure")
+
+hunter_autotools_configure_options(configure_opts
```
Comment (ruslo) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: Looks good for me :) We still have some boilerplate code (like `configure_command`/`build_command`/`ExternalProject_Add`) but it's not a big problem I guess.

Comment (pretyman) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: @ruslo I don't really know how to remove this boiler plate due to the use of `@VARIABLE@`. It makes me believe that  they won't be available as `${VARIABLE}`, when used with the `@`.

Let me know if you have any suggestions

File: cmake/find/Findodb-compiler.cmake (hunk 11 above)
Comment (pretyman) [inline cmake/find/Findodb-compiler.cmake]: Ok, I will look into the `install` and how can it be applied here.

FYI: this additional code is also needed by Findodb.cmake to find the libraries inside hunter's structure for packages. 
I am not sure if this is related but the Android toolchain has this line: https://github.com/ruslo/polly/blob/master/os/android.toolchain.cmake#L1532 which was the motivation for me adding this extra code. After adding this extra code, it managed to find the library and the compiler. I tested without the extra code and they were not being found.

File: cmake/find/Findodb.cmake (hunk 13 above)
Comment (pretyman) [inline cmake/find/Findodb.cmake]: Yes, I believe you are right, it is hardcoding the path somewhere, because it is failing at "./configure"  stage of odb-sqlite, with an error like:
dyld: Library not loaded: /home/myhome/.../.hunter/.../Build/.../libodb.dylib
  Referenced from: /home/myhome/../.hunter/.../Build/.../conftest
  Reason: image not found
Trace/BPT trap:5

(sorry I don't have the exact message with me right now)

Maybe it was something to do with RPATH, I have to read it up to understand it better.

I will separate the example at make a pull request at the end then, since the example relies in the compiler and `odb` package by itself does nothing, and was there to test `hunter_autotools_*`.

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in (hunk 15 above)
Comment (pretyman) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: Thanks, will do

File: cmake/find/Findodb-compiler.cmake (hunk 12 above)
Comment (pretyman) [inline cmake/find/Findodb-compiler.cmake]: I couldn't figure out how to `hunter_status_debug` here, since the list needs to be processed before adding it to the message

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in (hunk 16 above)
Comment (ruslo) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: @pretyman Yes, you're right. Variables will not be visible, you have to forward them manually like `set(FOO @FOO@)`. Well as I said before it looks good for me so probably you don't need bother with it.

File: cmake/find/Findodb-compiler.cmake (hunk 11 above)
Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: > the Android toolchain has this line

Hunter has this also: https://github.com/ruslo/hunter/blob/f711af744ade76be830307656dbded3c4cf244c4/cmake/Hunter#L73

It means adding `<root-id>/Install` to `CMAKE_FIND_ROOT_PATH`.

File: cmake/find/Findodb.cmake (hunk 13 above)
Comment (ruslo) [inline cmake/find/Findodb.cmake]: > Maybe it was something to do with RPATH, I have to read it up to understand better

Probably yes. See http://www.cmake.org/Wiki/CMake_RPATH_handling. The good news is that we can fix it without modifying native build instructions. We just need to add extra step with `install_name_tool` command which will fix RPATH of library/executable so package can be relocatable.

File: cmake/find/Findodb-compiler.cmake (hunk 12 above)
Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: You can keep code with creating the list and just replace `message(STATUS ...)` with `hunter_status_debug`

File: cmake/find/Findodb-compiler.cmake (hunk 11 above)
Comment (pretyman) [inline cmake/find/Findodb-compiler.cmake]: I looked up `install`, did a test project, but I could not work out how will I take a compressed archive which contains binaries in a structure like 
```
bin
doc
etc
lib
man
```

into `<root-id>/Install`. You can give me high level instructions about what you mean and I will search around on how to get it done. What I tried was to have the file uncompressed (with `ExternalProject_Add`, as a test) and then try to have `file(GLOB ...)` on the the contents of the source directory into a directory I specify, but it fails with "
```
CMake Error at CMakeLists.txt install PROGRAMS given directory 
"/Users/.../CMakeInstallTest/bin"
```
So it didn't look like I was on the right track.

Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: @pretyman take a look at this file: https://github.com/ruslo/hunter/blob/master/cmake/schemes/url_sha1_unpack.cmake.in probably just changing `SOURCE_DIR "@HUNTER_PACKAGE_SOURCE_DIR@"` to `SOURCE_DIR "@HUNTER_PACKAGE_INSTALL_PREFIX@"` will work

> then try to have file(GLOB ...)

Also I think you can't use `file(GLOB ...)` since there is no files on configure step (archive downloaded on build step). I.e. you need to write a CMake script `myscript.cmake` with `file(GLOB ...)` and add it to some `ExternalProject_Add` stage like `INSTALL_COMMAND "${CMAKE_COMMAND}" -P /path/to/mscript.cmake`

Comment (ruslo) [inline cmake/find/Findodb-compiler.cmake]: > just changing SOURCE_DIR "@HUNTER_PACKAGE_SOURCE_DIR@" to SOURCE_DIR "@HUNTER_PACKAGE_INSTALL_PREFIX@" will work

@pretyman though it may be not a good idea in case if this will remove existing files in directory. Please test it on small example

Comment (pretyman) [inline cmake/find/Findodb-compiler.cmake]: Hi Ruslo,
Indeed it 
>just changing SOURCE_DIR "@HUNTER_PACKAGE_SOURCE_DIR@" to SOURCE_DIR "@HUNTER_PACKAGE_INSTALL_PREFIX@" 
removes all contents of `@HUNTER_PACKAGE_INSTALL_PREFIX@`.

What I ended up doing was: create a `url_sha1_unpack_install.cmake.in` that uncompresses the tarball with `ExternalProject_Add` and then calls itself with `INSTALL_COMMAND` with a flag to `RUN_INSTALL`. Then I treat this flag with an `if` branching the behavior to execute a `file(COPY ...)` from the source to the `<root-id>/Install`. 

This seemed reasonable as I could keep the whole logic inside a single script.

It is working, I shall issue the pull request this weekend.

Thanks for all suggestions and sorry for the delay in getting back, but I was moving houses and been busy with other things.

File: cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in (hunk 15 above)
Comment (pretyman) [inline cmake/schemes/url_sha1_odb-sqlite_autotools.cmake.in]: This suggestion of yours, together with the comment that you mention that Hunter add itself to `CMAKE_FIND_ROOT` at https://github.com/ruslo/hunter/blob/f711af744ade76be830307656dbded3c4cf244c4/cmake/Hunter#L73 made it not necessary for me to add `ODB_ROOT` manually as I was doing inside the `if(ANDROID) ...` at https://github.com/ruslo/hunter/pull/175/files#diff-1c166191064e9c15bd530aeae37bb9b7R97

Thanks, working perfectly now!

File: cmake/find/FindSQLite3.cmake (hunk 10 above)
Comment (pretyman) [inline cmake/find/FindSQLite3.cmake]: Sharp eye, well spotted, thanks ;-)

File: cmake/find/Findodb.cmake (hunk 13 above)
Comment (pretyman) [inline cmake/find/Findodb.cmake]: I saved your comments in my notes and will close this pull request to open a new one without the examples, as discussed previously, then I will look into the RPATH issue.

**Comment by pretyman** [comment]
Closing this pull request to open a smaller one.
